scripts: size_report: Add support for TF-M and BL2 image size reports

joerchan opened the pull request:
Add support for TF-M and BL2 image size reports.
This adds the following targets when TF-M or BL2 is enabled:
tfm_rom_report, tfm_ram_report, tfm_footprint
bl2_rom_report, bl2_ram_report, bl2_footprint

Example:
west build -t tfm_rom_report

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
scripts: size_report: Add support for TF-M and BL2 image size reports

Add support for TF-M and BL2 image size reports.
This adds the following targets when TF-M or BL2 is enabled:
tfm_rom_report, tfm_ram_report, tfm_footprint
bl2_rom_report, bl2_ram_report, bl2_footprint

Example:
west build -t tfm_rom_report

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@carlescufi @nashif Any idea on where this could be documented?

microbuilder commented:
Looks good. Tested with:

```bash
$ west build -p -b mps2_an521_ns zephyr/samples/tfm_integration/tfm_ipc
$ west build -t tfm_rom_report
```

microbuilder commented:
> @carlescufi @nashif Any idea on where this could be documented?

Could we at least mention in the commit message that this commit adds support for:

```
$ west build -t tfm_rom_report
$ west build -t tfm_ram_report 
$ west build -t tfm_footprint
```

joerchan commented:
> Could we at least mention in the commit message that this commit adds support for:

I've updated the commit message, let me know if that looks good to you.

tejlmand reviewed (changes requested):
Could we please get this documented, perhaps here: https://docs.zephyrproject.org/latest/guides/tfm/build.html

Or a new optimization section for TF-M, similar to what we have for Zephyr itself.
For inspiration:
https://docs.zephyrproject.org/latest/guides/optimizations/tools.html#optimization-tools

joerchan committed:
doc: Fix whitespace alignment in example report

Fix whitespace alignment in example report.
This shows up unaligned in the published documentation.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
doc: tfm: Document size report commands for TF-M and BL2 images

Add documentation with example commands for getting the RAM and ROM
reports for the secure firmware image (TF-M) and mcuboot (BL2).

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
doc: tfm: Improve documentation of non-secure firmware image

Explain when the TF-M nonsecure firmware image is used.
Use non-secure consistently instead of nonsecure or ns.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@tejlmand I've added some examples on the commands and refer to the optimizations page for more information.

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -118,3 +118,42 @@ following CMake snippet in your CMakeLists.txt file.
+    :board: mps2_an521_ns
+    :goals: tfm_rom_report
+
+Use the ``bl2_ram_report`` to get the RAM report for mcuboot, if enabled.
```
just to be very precise that this is only the MCUboot built as part of TF-M.
```suggestion
Use the ``bl2_ram_report`` to get the RAM report for TF-M enabled MCUboot.
```

joerchan replied:
The " if enabled " part is just to say that this command is not available when TFM_BL2=n.

Do you wish to separate BL2 from MCUboot in the terminology?
I was aligning with the following text:
```
The TF-M build system creates the following executable files:
* tfm_s - the secure firmware
* tfm_ns - a nonsecure app which is discarded in favor of the Zephyr app
* bl2 - mcuboot, if enabled
```

Should we update to use TF-M MCUboot (or TF-M enabled MCUboot)?

microbuilder replied:
That's a fair point ... we should perhaps update this reference as well to be more precise and state: `bl2 - TF-M MCUBoot, if enabled`.

joerchan replied:
@microbuilder Thanks for the feedback, updated accordingly. Also changed to `TF-M secure firmware`

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -118,3 +118,42 @@ following CMake snippet in your CMakeLists.txt file.
+    :board: mps2_an521_ns
+    :goals: bl2_ram_report
+
+Use the ``bl2_rom_report`` to get the ROM report for mcuboot, if enabled.
```
similar here, I think it is important to make it clear to the reader that this is not just regular MCUboot, but TF-M build MCUboot.
```suggestion
Use the ``bl2_rom_report`` to get the ROM report for TF-M enabled MCUboot.
```

tejlmand reviewed (commented):
Thanks for adding docs.

Some small proposals.

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -118,3 +118,42 @@ following CMake snippet in your CMakeLists.txt file.
+    :board: mps2_an521_ns
+    :goals: tfm_rom_report
+
+Use the ``bl2_ram_report`` to get the RAM report for TF-M MCUBoot, if enabled.
```
Lower case `b` in MCUboot is the official writing:
```suggestion
Use the ``bl2_ram_report`` to get the RAM report for TF-M MCUboot, if enabled.
```

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -118,3 +118,42 @@ following CMake snippet in your CMakeLists.txt file.
+    :board: mps2_an521_ns
+    :goals: bl2_ram_report
+
+Use the ``bl2_rom_report`` to get the ROM report for TF-M MCUBoot, if enabled.
```
Lower case `b` in MCUboot is the official writing:
```suggestion
Use the ``bl2_rom_report`` to get the ROM report for TF-M MCUboot, if enabled.
```

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -22,9 +22,9 @@ Images Created by the TF-M Build
+* tfm_s - TF-M secure firmware
 * tfm_ns - a nonsecure app which is discarded in favor of the Zephyr app
-* bl2 - mcuboot, if enabled
+* bl2 - TF-M MCUBoot, if enabled
```
```suggestion
* bl2 - TF-M MCUboot, if enabled
```

tejlmand reviewed (dismissed):
Thanks for the last update.

LGTM, one minor nit noticed, `MCUboot` with lower case `b` is the official name.

tejlmand reviewed (dismissed):
lgtm - re-approved

joerchan commented:
@microbuilder @tejlmand Added another commit to clean up some of the non-secure documentation as well.

tejlmand commented on doc/guides/tfm/build.rst:
```diff
@@ -23,7 +23,7 @@ Images Created by the TF-M Build
 
 * tfm_s - TF-M secure firmware
-* tfm_ns - a nonsecure app which is discarded in favor of the Zephyr app
+* tfm_ns - TF-M non-secure app (only used by regression tests).
```
Thanks for pointing this out.

If we only use this for integration tests, wouldn't it be good to introduce a `CONFIG_TFM_NS` to control build of the TF-M ns app, similar to `CONFIG_TFM_BL2` ?
Then integration tests can enable building of TF-M ns when needed.

I see no reason for building the TF-M ns everytime if that is not needed.

If agreeing with this, please create an enhancement task for this work, as I believe  such cleanup to be outside scope of this PR.

joerchan replied:
Yes, that is a good point. I'll create an internal task for this.

There is a different Kconfig that determines which application is flashed, but it is tied to the regression tests in naming.

joerchan replied:
Follow up: https://github.com/zephyrproject-rtos/zephyr/pull/43599

tejlmand reviewed (approved):
lgtm.

I believe we should create enhancement task for improving TF-M build by disabling ns building when it's not needed, but that is orthogonal to this PR.
TF-M has an option to do so:
https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/5d32c3e64b3d589548e881eeeeb37d84944c90af/config/config_default.cmake#L14
```
set(NS ON CACHE BOOL "Whether to build NS app")
```
